torch/csrc/LazyTensor.cpp

```diff
@@ -0,0 +1,385 @@
+#ifndef C10_MOBILE
+
+#include <ATen/core/LazyTensor.h>
+#include <torch/csrc/jit/graph_executor.h>
+#include <torch/csrc/jit/interpreter.h>
+#include <torch/csrc/jit/operator.h>
+#include <torch/csrc/jit/script/function_schema_parser.h>
+
+
+/* LazyTensor Design
+ *
+ *
+ * LazyTensor is an abstraction that enables lazy execution in PyTorch.  It does so by
+ * providing a virtual tensor implementation that looks and feels like a regular Tensor
+ * but does not necessarily execute any of the underlying computation immediately.
+ *
+ * Instead of executing the computation "eagerly,"  LazyTensors will record requested
+ * operations into an implicit graph and later lower that graph to TorchScript IR.
+ * Once `to_eager()` is called, it is guaranteed that any computation required to
+ * resolve the tensor has been completed.
+ *
+ * For backends that digest TorchScript IR, this design is interesting because it enables
+ * these backends to be used in PyTorch without the need for a TorchScript frontend.
+ *
+ * Note: shape propagation is not supported! This means `to_eager` is required before
+ * any shape dependent code is run.  The implication is that some core modules,
+ * such as nn.BatchNorm, cannot be executed lazily because their implementations
+ * check shapes in Python.
```

Comment:
I feel in a discussion about a limitation here, it's important to distinguish whether or not a limitation is fundamental, or incidental. In this case, the limitation is incidental, because in principle we could have standalone shape inference for every function in which case BatchNorm would work fine.